8364970: Redo JDK-8327381 by updating the CmpU type instead of the Bool type

franferrax opened the pull request:
Hi, this pull request is a second take of 1383fec41756322bf2832c55633e46395b937b40, by updating the `CmpUNode` type as either `TypeInt::CC_LE` (case 1a) or `TypeInt::CC_LT` (case 1b) instead of updating the `BoolNode` type as `TypeInt::ONE`.

With this approach a56cd371a2c497e4323756f8b8a08a0bba059bf2 becomes unnecessary. Additionally, having the right type in `CmpUNode` could potentially enable further optimizations.

#### Testing

In order to evaluate the changes, the following testing has been performed:

* `jdk:tier1` (see [GitHub Actions run](https://github.com/franferrax/jdk/actions/runs/16789994433))
* [`TestBoolNodeGVN.java`](https://github.com/openjdk/jdk/blob/jdk-26+9/test/hotspot/jtreg/compiler/c2/gvn/TestBoolNodeGVN.java), created for [JDK-8327381: Refactor type-improving transformations in BoolNode::Ideal to BoolNode::Value](https://bugs.openjdk.org/browse/JDK-8327381) (1383fec41756322bf2832c55633e46395b937b40)
    * I also checked it breaks if I remove the `CmpUNode::Value_cmpu_and_mask` call
* Private reproducer for [JDK-8349584: Improve compiler processing](https://bugs.openjdk.org/browse/JDK-8349584) (a56cd371a2c497e4323756f8b8a08a0bba059bf2)
* A local run of the `test/hotspot/jtreg/compiler/c2` category on _Fedora Linux x86_64_
    * 🚧 ongoing

chhagedorn commented on src/hotspot/share/opto/subnode.cpp:
```diff
@@ -801,13 +801,92 @@ const Type *CmpUNode::sub( const Type *t1, const Type *t2 ) const {
+//
+// (1a) and (1b) is covered by this method since we can directly return the corresponding TypeInt::CC_*
+// while (2) is covered in BoolNode::Ideal since we create a new non-constant node (see [CMPU_MASK]).
+const Type* CmpUNode::Value_cmpu_and_mask(PhaseValues* phase, const Node* in1, const Node* in2) {
```
I suggest to directly name these:
`in1` -> `andI`
`in2`- > `rhs`

Then it's easier to follow the comments.

chhagedorn commented on src/hotspot/share/opto/subnode.cpp:
```diff
@@ -1814,100 +1893,10 @@ Node *BoolNode::Ideal(PhaseGVN *phase, bool can_reshape) {
-  }
-
-  return _test.cc2logical(input_type);
+  return _test.cc2logical( phase->type( in(1) ) );
```
```suggestion
  return _test.cc2logical(phase->type(in(1)));
```

franferrax replied:
Suggestion accepted in 27ed1a311ec34e24afae6cc43d2c71e2620eb0ef.

chhagedorn commented on src/hotspot/share/opto/phaseX.cpp:
```diff
@@ -2925,57 +2924,6 @@ void PhaseCCP::push_cmpu(Unique_Node_List& worklist, const Node* use) const {
-//        |
-//       Bool
-//
-void PhaseCCP::push_bool_with_cmpu_and_mask(Unique_Node_List& worklist, const Node* use) const {
```
Needed to double-check but I think it's fine to remove the notification code since we already have `push_cmpu()` in place which looks through the `AddI`:
https://github.com/openjdk/jdk/blob/10762d408bba9ce0945100847a8674e7eb7fa75e/src/hotspot/share/opto/phaseX.cpp#L2911-L2926

So, whenever `m` or `1` changes, we will re-add the `CmpU` to the CCP worklist with `push_cmpu()`. The `x` does not matter for the application of `Value_cmpu_and_mask()`.

franferrax replied:
Hmm, I was oversimplifying the problem, my way of thinking it was the following one:

```
m    x  m    1
 \  /    \  /
 AndI    AddI       grandparents
    \    /
     CmpU              parent
      |
     Bool            grandchild
```

_"As we were updating a grandchild based on its grandparents, we needed an ad-hoc worklist push for the grandchild. Since we now update the type of `CmpU` based on its parents, the canonical parent-to-child propagations should work, and we don't need any ad-hoc grandparents-to-grandchild worklist push anymore."_

But as you noted, non-immediate `CmpU` inputs such as `m` or `1` can change and should affect the `CmpU` type. Luckily, this already was the case for previous `CmpU` optimizations.

---

For case **1a**, we don't need `PhaseCCP::push_cmpu` because `m` is also an immediate input of `CmpU`.

```
m    x
 \  /
 AndI     m
    \    /
     CmpU
      |
     Bool
```

---

I'm now realizing this was a very lucky situation. The `AndI` input isn't problematic even when `PhaseCCP::push_cmpu` doesn't handle the `use_op == Op_AndI` case, because:

* `x` does not affect the application of `Value_cmpu_and_mask()`
* In case **1a**, `m` is a direct input of `CmpU`
* In case **1b**, the `AddI` input is handled in `PhaseCCP::push_cmpu` (`use_op == Op_AddI`)

Please let me know if you think we should add a comment in the code.

chhagedorn replied:
That's a good summary! Thanks for double-checking again. It's indeed only for **1b** a probably that's handled by `push_cmpu()`. It probably would not hurt to add a comment that `push_cmpu` handles this case, just to be sure.

chhagedorn commented on src/hotspot/share/opto/subnode.hpp:
```diff
@@ -171,6 +171,7 @@ class CmpINode : public CmpNode {
 //------------------------------CmpUNode---------------------------------------
 // Compare 2 unsigned values (integer or pointer), returning condition codes (-1, 0 or 1).
 class CmpUNode : public CmpNode {
+  static const Type* Value_cmpu_and_mask(PhaseValues*, const Node*, const Node*);
```
We usually add matching parameter names as found in the source file:
```
  static const Type* Value_cmpu_and_mask(PhaseValues* phase, const Node* in1, const Node* in2);
```
or with the renaming above:
```
  static const Type* Value_cmpu_and_mask(PhaseValues* phase, const Node* andI, const Node* rhs);

```

chhagedorn reviewed (approved):
Thanks for improving this! I have some small suggestions, otherwise, it looks good to me!

> Additionally, having the right type in CmpUNode could potentially enable further optimizations.

Could you already find some examples, where this change gives us an improved IR? If so, you could also add it as IR test.

I'll also give this a spinning in our testing.

chhagedorn reviewed (approved):
Update looks good, thanks! I'll give run some testing and report back again.

> Could you already find some examples, where this change gives us an improved IR? If so, you could also add it as IR test.

Just double-checking, were you able to find such a test which now improves the IR with the better type info and `CmpU` while we could not with the old code? Otherwise, you could also file a follow-up RFE.

franferrax commented:
> Thank you @franferrax for catching and addressing the inconsistent state. I neglected that in my original PR. I think it would be beneficial to include [your tables](https://github.com/openjdk/jdk/pull/26666#issuecomment-3183659818) of the two cases in the comments too. Thank you for the hard work.

@tabjy thanks for the suggestion, I added the tables in 32a7940b8eefe56c3aa603be65da7b32981d2ab7.

franferrax commented:
Hi @chhagedorn,

I added the new tests in e6b1cb897d9c75b34744c7d24f72abcec9986b0b. One problem I'm facing is that I'm unable to generate `Bool` nodes with arbitrary `BoolTest` values. Even if I try the assert inversions I removed in 10e1e3f4f796d05dcd5c56bc2365d5d564d93952, C2 has preference for `BoolTest::ne`, `BoolTest::le` and `BoolTest::lt`. Instead of using `BoolTest::eq`, `BoolTest::gt` or `BoolTest::ge`, it swaps what is put in `IfTrue` and `IfFalse`.

Even if `javac` generates an `ifeq` and an `ifne` with the same inputs, instead of a single `CmpU` with two `Bool`s (`BoolTest::eq` and `BoolTest::ne`), I get a single `Bool` (`BoolTest::ne`) with two `If` (one of them swapping `IfTrue` with `IfFalse`). I guess this is some sort of canonicalization to enable further optimizations.

Do you know a way to influence the `Bool`'s `BoolTest` value? Or @rwestrel do you?

This means the following 8 cases are not really testing what they claim, but repeating other cases with `IfTrue` and `IfFalse` swapped:

* `testCase1aOptimizeAsFalseForGT[xm|mx]` (they should use `BoolTest::gt`, but use `BoolTest::le`)
* `testCase1bOptimizeAsFalseForEQ[xm|mx]` (they should use `BoolTest::eq`, but use `BoolTest::ne`)
* `testCase1bOptimizeAsFalseForGE[xm|mx]` (they should use `BoolTest::ge`, but use `BoolTest::lt`)
* `testCase1bOptimizeAsFalseForGT[xm|mx]` (they should use `BoolTest::gt`, but use `BoolTest::le`)

Even if we don't find a way to influence the `BoolTest`, the cases are still valid and can be kept (just in case the described behaviour changes).

franferrax commented:
Learning a bit more about the IR tests framework, I noticed `testCorrectness` isn't probably doing what we want.

It should execute the compiled versions of the following `@Test` methods:

```
testCase(1a|1b)(OptimizeAsTrue|OptimizeAsFalse)For(EQ|NE|LE|GE|LT|GT)(xm|mx)
```

But the `@Test` methods warmup only occurs for `TestFramework.run()` and in a different JVM process.

So directly invoking `testCorrectness` outside `TestFramework.run()` is executed in the parent JVM without any warmup.

I think I fixed this in e2f8a43ce1ed2861c506af787018c38ed8769fe3 by making `testCorrectness` a `@Run`ner of those `@Test` methods, but please confirm.

franferrax commented:
# Absence note

Today is the last day before a ~2 weeks vacation, so my next working day is Monday, September 1st.

Please feel free to keep giving feedback and/or reviews, and I will continue when I'm back.

Cheers,
Francisco

merykitty commented on src/hotspot/share/opto/subnode.cpp:
```diff
@@ -801,13 +801,125 @@ const Type *CmpUNode::sub( const Type *t1, const Type *t2 ) const {
+      rhs_m = rhs->in(1);
+      const TypeInt* rhs_m_type = phase->type(rhs_m)->isa_int();
+      // Exclude any case where m == -1 is possible.
+      if (rhs_m_type != nullptr && (rhs_m_type->_lo > -1 || rhs_m_type->_hi < -1)) {
```
Please use `!rhs_m_type->contains(-1)`

franferrax replied:
Accepted in 25aa9d7e27a74cf6ed917af2d328f5880fe84de5. Smoke-test check: builds and the IR test passes.

franferrax commented:
Hi @chhagedorn, thank you for the additional work and your insights. This is much appreciated from a learner perspective.

I didn't fully analyze the `Test.java` you provided yet, but wanted to check if you are aiming to include the missing IGVN notification code as part of this issue (and its corresponding test). Or are you working on an independent issue?

My availability will be limited as the October CPU approaches, but it will try to find some timeboxes to make `TestBoolNodeGVN.java` emit the right test cases for `gt`, `ge` , and `eq`.

chhagedorn commented:
> Hi @chhagedorn, thank you for the additional work and your insights. This is much appreciated from a learner perspective.

Sure, you're welcome :-)

> I didn't fully analyze the Test.java you provided yet, but wanted to check if you are aiming to include the missing IGVN notification code as part of this issue (and its corresponding test). Or are you working on an independent issue?

I think you could squeeze that in here as well. With mainline, you probably need a different notification code because we need to add the `Bool` node instead of the `CmpU` node. But with this patch, we only require the `CmpU`. So, I guess it's not worth to fix it separately only to update it again with this patch.

> My availability will be limited as the October CPU approaches, but it will try to find some timeboxes to make TestBoolNodeGVN.java emit the right test cases for gt, ge , and eq

Sounds good, no hurry. Thanks for taking another look to improve the test!

franferrax commented:
/open
